### PR TITLE
Do not show in url branch name if only one is available

### PR DESCRIPTION
**Pull request by lecunm**

### CODE REVIEW

File: app/controllers/MenuController.scala [hunk 1]
```diff
@@ -9,7 +9,8 @@ import services.MenuService
 
 
 @Api(value = "MenuController", produces = "application/json")
-class MenuController @Inject()(menuService: MenuService) extends InjectedController {
+class
```
Comment (notflorian) [inline app/controllers/MenuController.scala]: Revert this

File: app/services/MenuService.scala [hunk 2]
```diff
@@ -52,10 +56,18 @@ class MenuService @Inject()(hierarchyRepository: HierarchyRepository, projectRep
             }
           ))
         }
-        val directory : Option[Directory] = hierarchyNode.directoryPath.flatMap{ directoryPath =>
-          directoryRepository.findByPath(directoryPath)
+        val directory: Option[Directory] = hierarchyNode.directoryPath.flatMap { directoryPath =>
+          directoryRepository.findByPath(directoryPath).map(d => {
```
Comment (notflorian) [inline app/services/MenuService.scala]: useless {

File: app/services/MenuService.scala [hunk 3]
```diff
@@ -111,10 +123,19 @@ object MenuService {
...
+  def buildDirectoryTree(currentDirectory: Directory, directories: Seq[Directory], project: Project): Directory = {
     val (children, left) = directories.partition(d => isChild("/", currentDirectory.relativePath, d.relativePath))
 
-    currentDirectory.copy(children = children.map(d => buildDirectoryTree(d, left)))
+    currentDirectory.copy(children = children.map(d => buildDirectoryTree(d, left, project)), path = getCorrectedPath(currentDirectory.path,project))
+
+  }
+
+  def getCorrectedPath(path: String, project: Project): String = {
+    if (project.stableBranch == project.displayedBranches.getOrElse("")) {
```
Comment (notflorian) [inline app/services/MenuService.scala]: I don't understand this if condition. project.displayedBranches is a regex

File: app/services/PageService.scala [hunk 4]
```diff
@@ -77,7 +78,16 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
 
   def getLocalRepository(projectId: String, branch: String): String = s"$projectsRootDirectory$projectId/$branch/".fixPathSeparator
 
-  def getPagePath(projectId: String, branch: String, path: String, documentationRootPath: String): String = s"$projectId>$branch>${path.substring(path.indexOf(documentationRootPath) + documentationRootPath.length, path.length)}".replace(".md", "")
+  def getPagePath(projectId: String, branch: String, path: String, documentationRootPath: String): String = {
+    s"$projectId>${
```
Comment (notflorian) [inline app/services/PageService.scala]: Fix formating

File: app/services/PageService.scala [hunk 5]
```diff
@@ -90,12 +100,14 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
 
         val currentDirectory = directoryRepository.save(Directory(-1, name, meta.directory.flatMap(_.label).getOrElse(name), meta.directory.flatMap(_.description).getOrElse(name), order, relativePath, path, branch.id))
 
-        val pages = meta.directory.flatMap(_.pages).getOrElse(Seq()).zipWithIndex.flatMap { case (pagePath, index) =>
-          processPage(currentDirectory, localDirectoryPath, pagePath, index)
+        val pages = meta.directory.flatMap(_.pages).getOrElse(Seq()).zipWithIndex.flatMap {
+          case (pagePath, index) =>
```
Comment (notflorian) [inline app/services/PageService.scala]: fomat issue

File: app/services/PageService.scala [hunk 6]
```diff
@@ -90,12 +100,14 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
...
+            processPage(currentDirectory, localDirectoryPath, pagePath, index)
         }
 
-        val children = meta.directory.flatMap(_.children).getOrElse(Seq()).zipWithIndex.flatMap { case (childName, index) =>
-          processDirectory(branch, path + childName + "/", localDirectoryPath + "/" + childName, relativePath + childName + "/", index, isRoot = false)
+        val children = meta.directory.flatMap(_.children).getOrElse(Seq()).zipWithIndex.flatMap {
+          case (childName, index) =>
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: app/services/PageService.scala [hunk 7]
```diff
@@ -109,15 +121,18 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
   }
 
   def computePageFromPath(path: String, refresh: Boolean = false): Option[PageWithContent] = {
+    val pathWithBranch = projectRepository.findById(path.split(">")(0)).map(project => {
```
Comment (notflorian) [inline app/services/PageService.scala]: useless {

File: app/services/PageService.scala [hunk 8]
```diff
@@ -134,15 +149,17 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
     pageRepository.findByPathJoinProject(path) match {
       case Some(pageJoinProject) =>
 
-        pageJoinProject.page.markdown.map { markdown =>
-          splitMarkdown(s"$markdown\n$MarkdownEnd", path)
-        }.map { fragments =>
-          processPageFragments(fragments, pageJoinProject)
+        pageJoinProject.page.markdown.map {
+          markdown =>
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: app/services/PageService.scala [hunk 9]
```diff
@@ -134,15 +149,17 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
...
+            splitMarkdown(s"$markdown\n$MarkdownEnd", path)
+        }.map {
+          fragments =>
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: app/services/PageService.scala [hunk 10]
```diff
@@ -134,15 +149,17 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
...
+            processPageFragments(fragments, pageJoinProject)
         } match {
           case Some(fragments) => {
             logger.debug(s"Page computed : $path")
-            val page = PageWithContent(pageJoinProject.page, fragments)
-            replacePageInCache(path,page)
+            val page = PageWithContent(pageJoinProject.page.copy(path= getCorrectedPath(path,pageJoinProject.project)), fragments)
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: app/services/PageService.scala [hunk 11]
```diff
@@ -155,8 +172,9 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
     val pageFile = new File(localDirectoryPath + "/" + name + ".md")
     if (pageFile.exists()) {
 
-      val pageOption = readPageContent(pageFile, name).map { case (content, label, description) =>
-        pageRepository.save(Page(-1, name, label, description, order, Some(content), currentDirectory.relativePath + name, currentDirectory.path + name, currentDirectory.id))
+      val pageOption = readPageContent(pageFile, name).map {
+        case (content, label, description) =>
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: app/services/PageService.scala [hunk 12]
```diff
@@ -166,8 +184,9 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
   def processPage(projectId: String, branchName: String, page: Page, documentationRootPath: String): Option[Page] = {
     val pageFile = new File(getLocalRepository(projectId, branchName) + documentationRootPath + "/" + page.relativePath + ".md")
     if (pageFile.exists()) {
-      val pageOption = readPageContent(pageFile, page.name).map { case (content, label, description) =>
-        pageRepository.save(page.copy(markdown = Some(content), label = label, description = description))
+      val pageOption = readPageContent(pageFile, page.name).map {
+        case (content, label, description) =>
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: app/services/PageService.scala [hunk 13]
```diff
@@ -184,10 +203,13 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
 
   def readPageContent(page: File, name: String): Option[(String, String, String)] = {
 
-    Try(FileUtils.readFileToString(page, "UTF-8")).logError(s"Error while reading content page ${page.getPath}").toOption.map { content =>
-      (for {
-        meta <- findPageModule(content)
-      } yield (content, meta.label.getOrElse(name), meta.description.orElse(meta.label).getOrElse(name))).getOrElse((content, name, name))
+    Try(FileUtils.readFileToString(page, "UTF-8")).logError(s"Error while reading content page ${
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: app/services/PageService.scala [hunk 14]
```diff
@@ -204,11 +226,12 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
 
 
   def splitMarkdown(markdown: String, path: String = ""): Seq[PageFragmentUnderProcessing] = {
-    markdown.split('\n').foldLeft((Seq[PageFragmentUnderProcessing](), PageFragmentUnderProcessing())) { (fragmentsAndCurrent, line) =>
-      fragmentsAndCurrent match {
-        case (fragments, currentFragment) =>
-          processPageFragmentLine(line, path, fragments, currentFragment)
-      }
+    markdown.split('\n').foldLeft((Seq[PageFragmentUnderProcessing](), PageFragmentUnderProcessing())) {
+      (fragmentsAndCurrent, line) =>
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: app/services/PageService.scala [hunk 15]
```diff
@@ -300,7 +323,9 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
               case Some(rawMarkdown) =>
                 val images = findPageImagesWithRelativePath(rawMarkdown)
                 val references = findPageReferencesWithRelativePath(rawMarkdown)
-                val markdown = (images ++ references).fold(rawMarkdown)((acc, relativePath) => acc.replace(relativePath, s"$baseUrl/api/assets?path=${pageJoinProject.directory.path}$relativePath"))
+                val markdown = (images ++ references).fold(rawMarkdown)((acc, relativePath) => acc.replace(relativePath, s"$baseUrl/api/assets?path=${
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: app/services/PageService.scala [hunk 16]
```diff
@@ -336,8 +361,9 @@ class PageService @Inject()(config: Configuration, projectRepository: ProjectRep
     val filter = ProjectMenuItem(project.id, project.name, branchName, featureFilter, tagsFilter)
     val documentation = gherkinRepository.buildProjectGherkin(filter)
 
-    documentation.branches.find(_.name.equals(branchName)).flatMap { branch =>
-      branch.features.find(_.path.equals(featureFilter.getOrElse("")))
+    documentation.branches.find(_.name.equals(branchName)).flatMap {
+      branch =>
```
Comment (notflorian) [inline app/services/PageService.scala]: format issue

File: test/services/ProjectServiceTest.scala [hunk 17]
```diff
@@ -47,7 +48,7 @@ class ProjectServiceTest extends WordSpec with MustMatchers with BeforeAndAfter
   val featureBranchDirectory = projectService.getLocalRepository(project.id, featureBranch)
   val bugfixBranchDirectory = projectService.getLocalRepository(project.id, bugfixBranch)
   val masterBranch = Branch(1, project.stableBranch, isStable = true, project.id)
-
+  val targetDirectory = projectService.projectsRootDirectory
```
Comment (notflorian) [inline test/services/ProjectServiceTest.scala]: rename to projectsRootDirectory

File: app/controllers/DirectoryController.scala [hunk 18]
```diff
@@ -18,14 +19,17 @@ class DirectoryController @Inject()(branchRepository: BranchRepository, director
...
+      paramsBranchName <- params.lift(1)
       relativePath <- params.lift(2)
 
+      project <- projectRepository.findById(projectId)
+      branchName = if (paramsBranchName.equals("")) project.stableBranch else paramsBranchName
+
       branchId <- branchRepository.findByProjectIdAndName(projectId, branchName).map(_.id)
       directory <- directoryRepository.findByBranchIdAndRelativePath(branchId, relativePath)
-      pagesPaths = pageRepository.findAllByDirectoryId(directory.id)
+      pagesPaths = pageRepository.findAllByDirectoryId(directory.id).map(page => page.copy(path = getCorrectedPath(page.path, project)))
```
Comment (loriepisicchio) [inline app/controllers/DirectoryController.scala]: This logic should not be in the controller. You should instead call pageService.getAllPagePaths and in the service, do whatever is necessary to retrieve the page paths

File: test/features/documentation/page/show_a_page_without_branch_name_in_path.feature [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+      | .01. | suggestion | Suggestion system | Projects      | Project    |
+    And we have the following projects
+      | id            | name                    | repositoryUrl                                         | stableBranch | displayedBranches | featuresRootPath | documentationRootPath |
+      | suggestionsWS | Suggestions WebServices | target/remote/data/GetFeatures/library/suggestionsWS/ | master       | master            | test/features    | doc                   |
+    And the links between hierarchy nodes are
+      | projectId     | hierarchyId |
+      | suggestionsWS | .01.        |
+    And we have those branches in the database
+      | id | name   | isStable | projectId     |
+      | 1  | master | true     | suggestionsWS |
```
Comment (loriepisicchio) [inline test/features/documentation/page/show_a_page_without_branch_name_in_path.feature]: Add tests with several branches in database

**Pull request merged**